src/routers/expenseRouter.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const express = require('express');
+
+const router = express.Router();
+
+const expenseServise = require('../services/expenses.js');
+const userServise = require('../services/users.js');
+
+router.post('/', (req, res) => {
+  const {
+    userId,
+    title,
+  } = req.body;
+
+  if (!title || !userServise.exist(userId)) {
+    res.sendStatus(400);
+
+    return;
+  }
+
+  const newExpense = {
+    ...req.body,
+    id: Math.floor(Math.random() * 10),
+  };
+
+  expenseServise.add(newExpense);
+
+  res.statusCode = 201;
+
+  res.send(newExpense);
+});
+
+router.get('/', (req, res) => {
+  const expenses = expenseServise.getAll();
+
+  const {
+    userId,
+    category,
+    from,
+    to,
+  } = req.query;
+
+  const id = Number(userId);
+
+  if (typeof id !== 'number') {
+    res.sendStatus(400);
+
+    return;
+  }
+
+  res.statusCode = 200;
+
+  if (!expenses.length) {
+    res.send([]);
+
+    return;
+  }
+
+  if (from && to) {
+    const userExpenses = expenseServise.filter(
+      expense => expense.spentAt > from && expense.spentAt < to
+    );
+
+    res.send(userExpenses);
+
+    return;
+  }
+
+  if (userServise.findById(id)) {
+    let userExpenses = expenseServise.filter(
+      expense => expense.userId === id
+    );
+
+    if (category) {
+      userExpenses = userExpenses.filter(
+        expense => expense.category === category
+      );
+    }
+
+    res.send(userExpenses);
+
+    return;
+  }
+
+  res.send(expenseServise.getAll());
+});
+
+router.get('/:id', (req, res) => {
+  const expenseId = Number(req.params.id);
+
+  if (typeof expenseId !== 'number') {
+    res.sendStatus(400);
+
+    return;
+  }
+
+  const foundExpense = expenseServise.findById(expenseId);
+
+  if (!foundExpense) {
+    res.sendStatus(404);
+
+    return;
+  }
+
+  res.statusCode = 200;
+
+  res.send(foundExpense);
+});
+
+router.patch('/:id', (req, res) => {
+  const expenseId = Number(req.params.id);
+
+  if (typeof expenseId !== 'number') {
+    res.sendStatus(400);
+
+    return;
+  }
+
+  const foundExpense = expenseServise.findById(expenseId);
+
+  if (!foundExpense) {
+    res.sendStatus(404);
+
+    return;
+  }
+
+  expenseServise.update(foundExpense, req.body);
```

Comment:
can you rewrite the update: send just the expense id, and request body?

Reply:
what's the problem with this implementation? I find my expense, then I call update method to update it with certain data

Reply:
- you find expense just for updating it (so it needs just for the update): when you will add db, you use just id, so it will more easier for you
- if you will use the update method several times, you need each time to add duplicated code to find expense that you need to update